File: extras/day1.py
from itertools import islice, starmap, tee
from operator import sub

flip = lambda f: lambda x, y: f(y, x)
drop = lambda n, xs: islice(xs, n, None)
nwise = lambda n, xs: zip(*starmap(drop, enumerate(tee(xs, n))))
pos = lambda x: x > 0
increases = lambda xs: sum(map(pos, starmap(flip(sub), nwise(2, xs))))

numbers = [*map(int, open("data/day1.txt"))]
window_sum = map(sum, nwise(3, numbers))

print(increases(numbers))
print(increases(window_sum))

File: app/Day1.hs
module Day1 where


import Control.Monad
import Data.List


parseLines :: String -> [Int]
parseLines = (map read) . lines


countIncreases :: [Int] -> Int
countIncreases = length . filter (>0) . diffs
    where diffs = zipWith (flip (-)) <*> tail


rollingWindow :: Int -> [a] -> [[a]]
rollingWindow n = map (take n) . (takeWhile bigEnough) . tails 
    where bigEnough x = length x >= n


main :: IO ()
main = do
    numbers <- parseLines <$!> readFile "data/day1.txt"

    putStr "Number of increases: "
    putStrLn $ show $ countIncreases numbers

    putStr "Number of increases after a rolling sum: "
    let windowSum = map sum $ rollingWindow 3 numbers
    putStrLn $ show $ countIncreases windowSum
